add gcloud auth

shashankkoppar opened the pull request.

Typositoire commented on README.md:
```diff
@@ -25,26 +24,36 @@ resource_types:
-* `stable_repo`: *Optional* A `false` value will disable using a default Helm stable repo. Any other value will be used to Override default Helm stable repo URL <https://charts.helm.sh/stable>. Useful if running helm deploys without internet access.
-* `tracing_enabled`: *Optional.* Enable extremely verbose tracing for this resource. Useful when developing the resource itself. May allow secrets to be displayed. (Default: false)
-* `helm_setup_purge_all`: *Optional.* Delete and purge every helm release. Use with extreme caution. (Default: false)
+-   `gcloud_cluster_auth`: _Optional._ Set to true to use gcloud service account file for kubernetes cluster authentication.
```
Please make this a separated section just like Digital Ocean.

shashankkoppar replied:
updated it separate as Digitial ocean!

Typositoire commented on assets/common.sh:
```diff
@@ -163,7 +186,14 @@ setup_resource() {
 
   echo "Initializing kubectl..."
   if [ "$digitalocean" == "false" ]; then
-    setup_kubernetes $1 $2
+    gcloud_cluster_auth=$(jq -r '.source.gcloud_cluster_auth // "false"' < $1)
+    if [ "$gcloud_cluster_auth" = "true" ]; then
+      echo "Initializing kubectl access using gcloud service account file"
+      setup_gcp_kubernetes $1 $2
+    else
+      echo "Initializing kubectl using certificates"
+      setup_kubernetes $1 $2
+    fi
   elif [ "$do_cluster_id" != "false" ] && [ "$do_access_token" != "false" ]; then
     echo "Initializing digitalocean..."
     setup_doctl $1 $2
```
I'm sure there's a better way to check for this now that we have gcloud, DO and normal. If you feel like you can fix this go ahead otherwise I'll try to figure out something before merging.

shashankkoppar replied:
Sure yeah, maybe i can check if we can do this with a switch loop

Typositoire replied:
Sure let's give this a try.

shashankkoppar replied:
switch case seems not so feasible here, but we can improve if condition itself, will commit. u can check!

Typositoire commented on README.md:
```diff
@@ -147,3 +158,20 @@ jobs:
+  type: helm
+  source:
+    gcloud_cluster_auth: true
+    gcloud_service_account_key_file: _plain service account json file_
```
I'd like this to be able to be a file path and check if first the value of `gcloud_service_account_key` is a file then read it, if not assume it's the content then write it to disk like you're doing now.

shashankkoppar replied:
Yes sounds good! will make the changes

shashankkoppar replied:
have done changes, please check in recent commit

Typositoire commented on assets/common.sh:
```diff
@@ -67,6 +67,29 @@ setup_kubernetes() {
+  gcloud config set account ${gcloud_service_account_name}
+  gcloud config set project ${gcloud_project_name}
+  gcloud container clusters get-credentials ${gcloud_k8s_cluster_name} --zone ${gcloud_k8s_zone}
+  kubectl version
```
Why ?

shashankkoppar replied:
this is how we will be able to set kubectl context using gcloud

Typositoire replied:
I mean the `kubectl version` :P

shashankkoppar replied:
ah lol just to confirm if kubectl is working fine. can remove it 😄😄

shashankkoppar replied:
removed this! :D

Typositoire commented on Dockerfile:
```diff
@@ -1,12 +1,16 @@
 RUN curl -sL -o /usr/local/bin/kubectl https://storage.googleapis.com/kubernetes-release/release/v${KUBERNETES_VERSION}/bin/linux/amd64/kubectl; \
   chmod +x /usr/local/bin/kubectl
 
+RUN curl -sSL https://sdk.cloud.google.com | bash
```
Is there a way to lock a version here? Less surprises in the future.

shashankkoppar replied:
Yes, let me check! indeed using specifc version is better

shashankkoppar replied:
fixed it latest commit :)

Typositoire reviewed (changes requested):
We'll also need to bump the version to `1.17.0`

shashankkoppar commented:
```
selected worker: concourseci-worker-1
Initializing kubectl access using gcloud service account file
Activated service account credentials for: [concourse@iam.gserviceaccount.com]
Updated property [core/account].
Updated property [core/project].
Fetching cluster endpoint and auth data.
kubeconfig entry generated for test.
Initializing helm...
```

this is how it looks in console output, have tested with concourse pipeline, works fine.

Typositoire commented on README.md:
```diff
@@ -25,26 +24,38 @@ resource_types:
+
+-   `gcloud_cluster_auth`: _Optional._ Set to true to use gcloud service account file for kubernetes cluster authentication.
+
+-   `gcloud_service_account_key_file`: _Optional_ Manadatory if gcloud_cluster_auth is set to true. Pass gcloud service accon json contents as value or a file path containing service_account json.
```
```suggestion
-   `gcloud_service_account_key_file`: _Optional_ Manadatory if gcloud_cluster_auth is set to true. Pass gcloud service account json contents as value or a file path containing service_account json.
```

shashankkoppar replied:
fixed the typo :D

Typositoire commented on assets/common.sh:
```diff
@@ -160,13 +189,19 @@ setup_resource() {
+  if [ "$do_cluster_id" != "false" ] && [ "$do_access_token" != "false" ]; then
     echo "Initializing digitalocean..."
     setup_doctl $1 $2
+  else
+    if [ "$gcloud_cluster_auth" = "true" ]; then
+      echo "Initializing kubectl access using gcloud service account file"
+      setup_gcp_kubernetes $1 $2
+    else
+      echo "Initializing kubectl using certificates"
+      setup_kubernetes $1 $2
+    fi
```
```suggestion
  elif [ "$gcloud_cluster_auth" = "true" ]; then
      echo "Initializing kubectl access using gcloud service account file"
      setup_gcp_kubernetes $1 $2
  else
    echo "Initializing kubectl using certificates"
    setup_kubernetes $1 $2
  fi
```

shashankkoppar replied:
my bad, yes updated :)

Typositoire reviewed (changes requested):
Some small changes left but this looks good!

Typositoire commented:
I'll merge it as latest and tag it when I have 5 minutes my concourse pipeline for tagging doesn't seem to be working right now xD

shashankkoppar commented:
Thanks man, appreciate it!
